src/core/components/version.js

```diff
@@ -3,17 +3,24 @@
 const pkg = require('../../../package.json')
 const promisify = require('promisify-es6')
 
+// TODO add the commit hash of the current ipfs version to the response.
 module.exports = function version (self) {
   return promisify((opts, callback) => {
     if (typeof opts === 'function') {
       callback = opts
       opts = {}
     }
 
-    callback(null, {
-      version: pkg.version,
-      repo: '',
-      commit: ''
+    self.repo.version((err, repoVersion) => {
+      if (err) {
+        throw err
+      }
+
+      callback(null, {
+        version: pkg.version,
+        repo: repoVersion,
+        commit: ''
+      })
```

Comment:
@JonKrone found the issue. Here you are assuming that IPFS has always been init when it is not the case always. the Sharness test is checking the version with ipfs offline and without init.

Reply:
Thanks! In order to avoid the init constraint, go-ipfs accesses the repo version directly from a constant exposed by [go-ipfs/repo](https://github.com/ipfs/go-ipfs/blob/master/core/commands/version.go#L40) so the repo version is available offline and without an initialized repo. Analogous to `const repoVersion = require('ipfs-repo').version`

Unfortunately, there's no similar field exposed by the `js-ipfs-repo` package. We could read the [`js-ipfs-repo/constants`](https://github.com/ipfs/js-ipfs-repo/blob/master/src/constants.js) file directly like we're doing with `pkg.version` above or add the version as an export of `js-ipfs-repo`.

If we need to be able to read the version w/o init, I think an export is the way to go and replace the dependence of [`repo.version`](https://github.com/ipfs/js-ipfs/blob/master/src/core/components/repo.js#L10) on an initialized `_repo` with just `require('ipfs-repo').version`. Should we then add this version constant to the repo SPEC?

Feels like a big change to not use the repo API, what are your thoughts?

Reply:
We can totally expose the constant, however, I do not think is the best approach. If we use the version from the module, then we might have a false positive where the module is signaling the latest repo version than the one that is inited.

I feel that the right way to do it is:

- if repo inited, then read the `version` from the repo
- if the repo is not inited, then read the `version` from the constant

Reply:
Cool, I agree! I had also looked into what happens when the version is outdated but got a little caught up because starting the daemon with an outdated `~./jsipfs/version` throws during startup, making me think that we just bailed on mismatches. test: `echo "5" > ~/.jsipfs/version && jsipfs --help`

If that's expected behavior, I can't think of a case where you can run a cli command on an initialized but outdated repo, and we could safely report the constant `version`.

I think you have the right of it, though, and should aim for reporting the version of `~/.jsipfs/version` if the repo is initialized and maybe at some point in the future, change startup so that we delay opening the repo connection or make accessing the version independent of opening a repo?

either way: https://github.com/ipfs/js-ipfs-repo/pull/158